Python/specialize.c

```diff
@@ -1825,104 +1826,176 @@ function_get_version(PyObject *o, int opcode)
     return version;
 }
 
-void
-_Py_Specialize_BinarySubscr(
-     _PyStackRef container_st, _PyStackRef sub_st, _Py_CODEUNIT *instr)
+static int
+binop_subscr_list_int_guard(PyObject *container, PyObject *sub)
 {
-    PyObject *container = PyStackRef_AsPyObjectBorrow(container_st);
-    PyObject *sub = PyStackRef_AsPyObjectBorrow(sub_st);
+    if (!PyList_CheckExact(container) || !PyLong_CheckExact(sub)) {
+        return 0;
+    }
+    PyLongObject *index_obj = (PyLongObject *)sub;
+    if (!_PyLong_IsNonNegativeCompact(index_obj)) {
+        return 0;
+    }
+    PyListObject* list = (PyListObject*)container;
+    Py_ssize_t index = index_obj->long_value.ob_digit[0];
+    if (index >= PyList_GET_SIZE(list) || PyList_GET_ITEM(list, index) == NULL) {
+        return 0;
+    }
+    return 1;
+}
 
-    assert(ENABLE_SPECIALIZATION_FT);
-    assert(_PyOpcode_Caches[BINARY_SUBSCR] ==
-           INLINE_CACHE_ENTRIES_BINARY_SUBSCR);
-    PyTypeObject *container_type = Py_TYPE(container);
-    uint8_t specialized_op;
-    if (container_type == &PyList_Type) {
-        if (PyLong_CheckExact(sub)) {
-            if (_PyLong_IsNonNegativeCompact((PyLongObject *)sub)) {
-                specialized_op = BINARY_SUBSCR_LIST_INT;
-                goto success;
-            }
-            SPECIALIZATION_FAIL(BINARY_SUBSCR, SPEC_FAIL_OUT_OF_RANGE);
-            goto fail;
-        }
-        SPECIALIZATION_FAIL(BINARY_SUBSCR,
-            PySlice_Check(sub) ? SPEC_FAIL_SUBSCR_LIST_SLICE : SPEC_FAIL_OTHER);
-        goto fail;
+static PyObject*
+binop_subscr_list_int_action(PyObject *container, PyObject *sub)
+{
+    STAT_INC(BINARY_SUBSCR, hit);
+    Py_ssize_t index = ((PyLongObject*)sub)->long_value.ob_digit[0];
+    PyListObject* list = (PyListObject*)container;
+#ifdef Py_GIL_DISABLED
+    PyObject *res_o = _PyList_GetItemRef(list, index);
+    assert(res_o != NULL);
```

Comment:
What if the list is resized after the guard has been executed and before the action is executed. Then `_PyList_GetItemRef` can return NULL.